Improve implicit search drastically

jvican opened the pull request:
Disable logging of pickler generation
- Keep around the infrastructure used for debugging the implicit search
  because we'll come back to improve it further soon.

Add big improvement in the implicit search
- Change all the implicit definitions to return
  `AbstractPicklerUnpickler` instead of `Pickler[T] with Unpickler[T]`.
  The former one has more precedence than the latter one. This avoids to
  call the generator macro EVERY TIME we ask for an implicit
  pickler/unpickler.
- Make sure that the `LowPriorityImplicits` are placed in the right
  location of the `Defaults` trait.
- This addresses a big deal of #422 although more changes will come,
  especially affecting `preferringAlternativeImplicits`.
- Do some minor cleanup and formatting changes (trying out scalafmt) in
  the project.
- Remove explicit registerings of picklers that were not necessary
  because of `AutoRegister`.
- Register myself to the TODOs that are concerned with the creation of
  runtime picklers/unpicklers. @jsuereth, I've stolen you some TODOs ;)

Add benchmark utility that helps address #422
- Add just a memo that will tell us how many times we are looking for the
  pickler/unpickler of a type `T`.

Remove cumbersome `PicklingMacros` definition
- We don't need `PicklingMacros` when we can use the more explicit
  `Pickler.generate` and family.
- Rename `Macros` to `PicklingMacros`.

Move `preferringAlternativeImplicits` to Macros
- Reuse `AlgorithmLogger` to report errors
- Do some basic cleanup and lay the ground for future work on improving the
  `preferringAlternativeImplicits`.

First step towards better implicit search
- Improve debugging of the algorithm and put some comments.
- Gather `GenPicklers` and `GenUnpicklers` in the same file, remove
  unused trait. Update `AllPicklerUnpicklers` accordingly.

jvican committed:
First step towards better implicit search

* Improve debugging of the algorithm and put some comments.

* Gather `GenPicklers` and `GenUnpicklers` in the same file, remove
  unused trait. Update `AllPicklerUnpicklers` accordingly.

jvican committed:
Move `preferringAlternativeImplicits` to Macros

* Reuse `AlgorithmLogger` to report errors

* Do some basic cleanup and lay the ground for future work on improving
the `preferringAlternativeImplicits`.

jvican committed:
Remove cumbersome `PicklingMacros` definition

* We don't need `PicklingMacros` when we can use the more explicit
`Pickler.generate` and family.

* Rename `Macros` to `PicklingMacros`.

jvican committed:
Add benchmark utility that helps address #422

* Add just a memo that will tell us how many times we are looking for
the pickler/unpickler of a type `T`.

jvican committed:
Add big improvement in the implicit search

* Change all the implicit definitions to return
  `AbstractPicklerUnpickler` instead of `Pickler[T] with Unpickler[T]`.
  The former one has more precedence than the latter one. This avoids to
  call the generator macro EVERY TIME we ask for an implicit
  pickler/unpickler.

* Make sure that the `LowPriorityImplicits` are placed in the right
  location of the `Defaults` trait.

* This addresses a big deal of #422 although more changes will come,
  especially affecting `preferringAlternativeImplicits`.

* Do some minor cleanup and formatting changes (trying out scalafmt) in
  the project.

* Remove explicit registerings of picklers that were not necessary
  because of `AutoRegister`.

* Register myself to the TODOs that are concerned with the creation of
  runtime picklers/unpicklers. @jsuereth, I've stolen you some TODOs ;)

jvican committed:
Disable logging of pickler generation

* Keep around the infrastructure used for debugging the implicit search
  because we'll come back to improve it further soon.

jvican commented:
I'm still polishing a little bit this PR and make it pass!

jvican committed:
Add compat import to preferExistingImplicits

jsuereth commented:
FYI you should merge with your test fixes.  That was passing os I merged.  Looks like I have a bunch of merge conflicts to resovle in my branch too :)

jvican committed:
Fix compat error in 2.10 and wikigraph issues

* Import `HasCompat` to expose the 2.10 macros stub and move around the
  implicit class hack to make `pt` compile. This is quite flaky, we
  should use something like `macro-compat`.

* Change type signature of picklers/unpicklers in wikigraph to avoid
  clash.

jvican commented:
This is ready for review :+1: I've fixed the compatibility issues with 2.10.

jvican commented:
This is also ready for merge @jsuereth

jsuereth commented on core/src/main/scala/scala/pickling/Tools.scala:
```diff
@@ -353,45 +354,6 @@ abstract class Macro extends RichTypes { self =>
   def syntheticPicklerUnpicklerQualifiedName(tpe: Type): TypeName = syntheticBaseQualifiedName(tpe) + syntheticPicklerUnpicklerSuffix()
   def syntheticPicklerUnpicklerSuffix(): String = "PicklerUnpickler"
 
-  def preferringAlternativeImplicits(body: => Tree): Tree = {
```
Nice!

jsuereth commented on core/src/main/scala/scala/pickling/generator/PicklingMacros.scala:
```diff
@@ -0,0 +1,129 @@
+    val ourPt = candidates.head.pt
+
+    val key = ourPt.toString
+    Memo.memo += key -> (Memo.memo.getOrElse(key, 0) + 1)
```
Should the debugging code still be around?

jvican replied:
Let's leave it for a moment, I want to play around with the implicit search a little bit more. It's only used at compile-time, in the final release will be removed :+1:  Is this ok to be merged?

jvican replied:
I've updated this PR and removed this.

jvican committed:
Remove debugging infrastructure

jvican commented:
Closed and reopened to force Travis to execute the tests.
